xds/src/test/java/io/grpc/xds/ClientXdsClientV3Test.java

```diff
@@ -656,4 +666,118 @@ public boolean matches(LoadStatsRequest argument) {
+    Listener listener = ServerXdsClientNewServerApiTest.buildListenerWithFilterChain(
+        LISTENER_RESOURCE,
+        7000,
+        "0.0.0.0",
+        "google-sds-config-default",
+        "ROOTCA");
+    List<Any> listeners = ImmutableList.of(Any.pack(listener));
+    call.sendResponse(ResourceType.LDS, listeners, "0", "0000");
+    // Client sends an ACK LDS request.
+    call.verifyRequest(ResourceType.LDS, Collections.singletonList(LISTENER_RESOURCE), "0",
+        "0000", NODE);
+    verify(ldsResourceWatcher).onChanged(ldsUpdateCaptor.capture());
+    assertThat(ldsUpdateCaptor.getValue().listener)
+        .isEqualTo(EnvoyServerProtoData.Listener.fromEnvoyProtoListener(listener));
+
+    listener = ServerXdsClientNewServerApiTest.buildListenerWithFilterChain(
+        LISTENER_RESOURCE,
+        7000,
+        "0.0.0.0",
+        "CERT2",
+        "ROOTCA2");
+    listeners = ImmutableList.of(Any.pack(listener));
+    call.sendResponse(ResourceType.LDS, listeners, "1", "0001");
+
+    // Client sends an ACK LDS request.
+    call.verifyRequest(ResourceType.LDS, Collections.singletonList(LISTENER_RESOURCE), "1",
+        "0001", NODE);
+    verify(ldsResourceWatcher, times(2)).onChanged(ldsUpdateCaptor.capture());
+    assertThat(ldsUpdateCaptor.getValue().listener)
+        .isEqualTo(EnvoyServerProtoData.Listener.fromEnvoyProtoListener(listener));
+
+    assertThat(fakeClock.getPendingTasks(LDS_RESOURCE_FETCH_TIMEOUT_TASK_FILTER)).isEmpty();
+  }
+
+  @Override
+  public void serverSideListenerNotFound() {
+    ClientXdsClientTestBase.DiscoveryRpcCall call =
+        startResourceWatcher(LDS, LISTENER_RESOURCE, ldsResourceWatcher);
+    final FilterChain filterChainInbound = ServerXdsClientNewServerApiTest
+        .buildFilterChain(ServerXdsClientNewServerApiTest.buildFilterChainMatch("managed-mtls"),
+            CommonTlsContextTestsUtil.buildTestDownstreamTlsContext("google-sds-config-default",
+                "ROOTCA"),
+            ServerXdsClientNewServerApiTest.buildTestFilter("envoy.http_connection_manager"));
+    Listener listener = ServerXdsClientNewServerApiTest.buildListenerWithFilterChain(
+        "grpc/server?xds.resource.listening_address=0.0.0.0:8000",
+        7000,
+        "0.0.0.0",
+        filterChainInbound);
+    List<Any> listeners = ImmutableList.of(Any.pack(listener));
+    call.sendResponse(ResourceType.LDS, listeners, "0", "0000");
+    // Client sends an ACK LDS request.
+    call.verifyRequest(ResourceType.LDS, Collections.singletonList(LISTENER_RESOURCE), "0",
+        "0000", NODE);
+
+    verifyNoInteractions(ldsResourceWatcher);
+    fakeClock.forwardTime(ClientXdsClient.INITIAL_RESOURCE_FETCH_TIMEOUT_SEC, TimeUnit.SECONDS);
+    verify(ldsResourceWatcher).onResourceDoesNotExist(LISTENER_RESOURCE);
+    assertThat(fakeClock.getPendingTasks(LDS_RESOURCE_FETCH_TIMEOUT_TASK_FILTER)).isEmpty();
+  }
+
+  @Override
+  public void serverSideListenerNotInbound() {
```

Comment:
This test case is cumbersome to be tested with the client-server flow, and you are not verifying the NACK response's error message. The easiest way to cover such is to make your `parseServerSideListener()` a static helper method, and then write test cases to test that individually. See [ClientXdsClientDataTest](https://github.com/grpc/grpc-java/blob/master/xds/src/test/java/io/grpc/xds/ClientXdsClientDataTest.java).

Reply:
The existing `DiscoveryRequestMatcher` didnt include checking errorDetail and I have made that change in my next PR. Don't you think we should have at least one test to verify that a NACK Request from the client is verified correctly?

Reply:
There are too many branches for NACKing, unit test can never fully cover them. So, I wouldn't try to enumerate them. On the other hand, if we well organize our code and decently test message parsing/validation logics (static methods), it should be robust enough.

Reply:
replaced the test with a new test in `ClientXdsClientDataTest`